### ISSUE TITLE
Pyright incorrectly tries to use __get__ descriptor to resolve instance members

### ISSUE THREAD

**Issue by QuantumTim** [microsoft/pyright#9107]
**Describe the bug**
If you add an instance member (*not* a class member) that implements the descriptor protocol, then pyright will still try to use it as if it were a class member (ie it will incorrectly "call" `__get__` on it).

This behaviour seems to depend on whether you add a type annotation for the member, or whether it is inferred. Perhaps pyright is incorrectly treating the annotation as a class member rather than an instance member?

**Code or Screenshots**
```python
from __future__ import annotations
from typing import Any, Self, overload, reveal_type

class bound_method:
    _method: method # <-- If you remove this line, then it type-checks correctly.
    def __init__(self, m: method):
        self._method = m
    def get(self):
        return self._method

class method:
    @overload
    def __get__(self, instance: None, owner: type[Any]) -> Self: ...
    @overload
    def __get__(self, instance: Any, owner: type[Any]|None = None) -> bound_method: ...
    def __get__(self, instance: Any|None, owner: type[Any]|None = None) -> bound_method|Self:
        if instance is None:
            return self
        return bound_method(self)

class Foo:
    bar = method()

foo = Foo()
m = foo.bar.get()
reveal_type(Foo.bar) # Runtime type is 'method'
reveal_type(foo.bar) # Runtime type is 'bound_method'
reveal_type(m)       # Runtime type is 'method' <-- pyright infers 'bound_method'
```

**VS Code extension or command-line**
Using both the pyright VSCode plugin (v1.1.382), and pyright from PyPI on the command-line (1.1.382).

**Comment by erictraut** [comment]
Pyright is working as intended here. You're declaring `_method` as a variable that is a "regular class variable" — a variable that can be assigned as a class variable or shadowed as an instance variable. See [this section of pyright's documentation](https://microsoft.github.io/pyright/#/type-concepts-advanced?id=class-and-instance-variables) for details.

If you want to define the variable as a [pure instance variable](https://microsoft.github.io/pyright/#/type-concepts-advanced?id=pure-instance-variables), you should declare its type only within an instance method, like this:

```python
class bound_method:
    def __init__(self, m: method):
        self._method: method = m

    def get(self):
        return self._method
```

**Comment by QuantumTim** [comment]
Thanks for the quick response!

Hmm, okay. The Python docs don't make that clear either in the original [PEP 526](https://peps.python.org/pep-0526/#class-and-instance-variable-annotations) document, where it calls it an "instance variable without default", nor in the [current docs for ClassVar](https://docs.python.org/3/library/typing.html#typing.ClassVar), which just calls it an "instance variable". And I can't find anything in the [typing specs](https://typing.readthedocs.io/en/latest/spec/annotations.html) at all that refers to it either way. However, I have just tried with mypy, which seems to give similar results, so if this is the wider accepted practice (rather than something pyright-specific), perhaps some of the docs need updating. Happy to propose that over on the typing forum if that's the case.

I had actually come across your method as a work-around... however when I tried using it in my real code, it didn't work. Took me a while, but I just figured out why... apparently the presence of `__slots__` is what causes it to break. If you add `__slots__ = ('_method', )` then for some reason both pyright and mypy revert to using the descriptor and inferring it as `bound_method`. If anything I would have thought the presence of the slot would make it even clearer that this must be an instance-only member, since the internal slot descriptor object would be the class member of the same name. With the slots in, even if you annotate `self._method: method`, pyright doesn't complain about trying to access the class member `bound_method._method` (which it does complain about without slots), but it tells me its of type `method` instead of something like `slot_descriptor[method]` or whatever.

The other difference in my real code, is that I'm using `__new__` instead of `__init__`. If I try to annotate `self._method: method` in `__new__`, I get "error: Type annotation not supported for this statement" (or with mypy: "error: Type cannot be declared in assignment to non-self attribute"). Perhaps something could be added to support these annotations in a `__new__` function if the variable is called `self`? Or even just support inferring members and types inside `__new__`. I think the only reason I was trying to add an explicit annotation in the first place was because of the errors if you try assigning to stuff in `__new__`!

**Comment by erictraut** [comment]
> I can't find anything in the [typing specs](https://typing.readthedocs.io/en/latest/spec/annotations.html) at all that refers to it either way.

Yeah, this is an area where the typing spec isn't currently very clear. We've made good progress over the past year clarifying many previously-ambiguous parts of the spec. but this is an area that still requires some work. [Here](https://docs.google.com/presentation/d/1ash1ellodiV6FTyVnY-Q6V2G5vJ68ycWBr6MYkayC1U/edit#slide=id.g2770f948faa_0_5) is a slide from a presentation I gave several months ago at a meetup of typing community members. You can see that "methods & attributes" features prominently in this slide. I'm hoping that we can augment this part of the spec over the next year. If and when the spec becomes clear, pyright will be modified (if necessary) to [conform](https://htmlpreview.github.io/?https://github.com/python/typing/blob/main/conformance/results/results.html) to the specified behavior.

> Perhaps something could be added to support these annotations in a __new__ function if the variable is called `self`?

Type rules based on variable names wouldn't be sound. Variable names are a convention only, not part of the language spec.

> Or even just support inferring members and types inside `__new__`. 

That's unlikely. The intent would be ambiguous. Plus, I'd argue that the `__new__` method isn't the appropriate place to be declaring (or even assigning) instance variables. 

Pyright allows you to declare the types of pure instance variables within any instance method. It doesn't need to be an `__init__` method. For example, you could add the declaration to your `get` method.

```python
class bound_method:
    def __init__(self, m: method):
        self._method = m

    def get(self):
        self._method: method
        return self._method
```

You can also omit the declaration, and pyright (and mypy) will correctly infer its type and the fact that it's a pure instance variable based on its usage.

```python
class bound_method:
    def __init__(self, m: method):
        self._method = m

    def get(self):
        return self._method
```

**Comment by QuantumTim** [comment]
I think I can probably switch to `__init__`, which would solve most of the problems. I think I was using `__new__` because some other code needed it for a case that was more complicated and it didn't want to work with `__init__`, and then I got paranoid and used `__new__` in a few more places that probably didn't really need it :sweat_smile:.

Using `__slots__` still causes an issue though (even with `__init__`):

```python
class bound_method:
    __slots__ = ('_method', )
    def __init__(self, m: method) -> None:
        self._method = m
        reveal_type(self._method) # pyright: Type of "self._method" is "method"
    def get(self):
        reveal_type(self._method) # pyright: Type of "self._method" is "bound_method"
        return self._method
```
